README.md
# tacc
Total AccountAbility/Customer Care (TACC) is an Internet Service Provider (ISP) billing system written in C++ utilizing the Qt toolkit.


doc/ChangeLog.txt
------------------------------------------------------------------------
r196 | marc | 2009-03-20 12:30:03 -0700 (Fri, 20 Mar 2009) | 2 lines

Credit card payments can now be made against previously used cards.
Added a ConfirmBox to credit card payments.
------------------------------------------------------------------------
r195 | marc | 2009-03-20 11:22:42 -0700 (Fri, 20 Mar 2009) | 2 lines

Added new CCTools utilities.
Updated CCRefund to use card list generated from CCTools.
------------------------------------------------------------------------
r194 | marc | 2009-03-19 15:40:13 -0700 (Thu, 19 Mar 2009) | 1 line

Minor change to seanetimport to accomodate null dates for 'accountOpened' and 'nextBillDate'
------------------------------------------------------------------------
r193 | marc | 2009-03-19 14:52:36 -0700 (Thu, 19 Mar 2009) | 1 line

Fixed the file permission problems in the spec file for real this time.
------------------------------------------------------------------------
r192 | marc | 2009-03-19 13:19:44 -0700 (Thu, 19 Mar 2009) | 1 line

Minor spec filie tweaks to fix permissions
------------------------------------------------------------------------
r191 | marc | 2009-03-19 12:19:22 -0700 (Thu, 19 Mar 2009) | 1 line

Added RPM creation scripts
------------------------------------------------------------------------
r190 | marc | 2009-03-19 11:38:10 -0700 (Thu, 19 Mar 2009) | 1 line

Removed dependancy for external key from ccget
------------------------------------------------------------------------
r189 | marc | 2009-03-18 15:48:12 -0700 (Wed, 18 Mar 2009) | 2 lines

Fixed FS#9 - Typo in customer Payments menu.
Added wait cursor and status notification to General Ledger Detail Report
------------------------------------------------------------------------
r188 | marc | 2009-03-18 12:23:04 -0700 (Wed, 18 Mar 2009) | 2 lines

New ConfirmBox class added for extra verification.
Updated CCRefund to use the new ConfirmBox to verify the transaction
------------------------------------------------------------------------
r187 | marc | 2009-03-18 10:48:22 -0700 (Wed, 18 Mar 2009) | 1 line

CCRefund will now handle AutoPayment cards as well as those from previous transactions.  Basically any card that a customer has ever paid with.
------------------------------------------------------------------------
r186 | marc | 2009-03-17 16:56:12 -0700 (Tue, 17 Mar 2009) | 1 line

Credit Card Refunds now functional
------------------------------------------------------------------------
r185 | marc | 2009-03-17 11:23:49 -0700 (Tue, 17 Mar 2009) | 1 line

Updated credit card payment charge day to a spinbox.
------------------------------------------------------------------------
r184 | marc | 2009-03-16 17:11:16 -0700 (Mon, 16 Mar 2009) | 1 line

Fixed statement logic for anniversary billing
------------------------------------------------------------------------
r183 | marc | 2009-03-16 15:43:15 -0700 (Mon, 16 Mar 2009) | 2 lines

Can now search by RegNum.
Added clipboard copy.  Will copy FullName, ContactName, the currently selected PhoneNumber and RegNum to the clipboard.
------------------------------------------------------------------------
r182 | marc | 2009-03-16 13:21:25 -0700 (Mon, 16 Mar 2009) | 4 lines

Fixed bug in tmpFileName.
Updated RunSubscriptions class to use the correct columns in Accounts.
Updated seanetimport to use 'Yesterday' for balance transfers.
Cleaned up temp file handling in emailLatexStatements
------------------------------------------------------------------------
r181 | marc | 2009-03-16 09:57:47 -0700 (Mon, 16 Mar 2009) | 1 line

Updated BillingSettings and CCMonetra to allow the credit card receipts and declined messages to be optional.
------------------------------------------------------------------------
r180 | marc | 2009-03-15 17:35:11 -0700 (Sun, 15 Mar 2009) | 3 lines

Added new EmailTemplatesSettings class to configure credit card notices.
Updated CCBatch and CCMonetra to use new configurable email template names.
Updated CCBatch and CCMonetra to email to statement contacts instead of primary login.
------------------------------------------------------------------------
r179 | marc | 2009-03-15 15:26:23 -0700 (Sun, 15 Mar 2009) | 1 line

Added an email template editor, updated CCMonetra.cpp to use config values for domains instead of the hardcoded ones
------------------------------------------------------------------------
r178 | marc | 2009-03-15 10:51:52 -0700 (Sun, 15 Mar 2009) | 1 line

Minor fix for customer balance if doing historical data in seanetimport.cpp
------------------------------------------------------------------------
r177 | marc | 2009-03-15 09:18:35 -0700 (Sun, 15 Mar 2009) | 1 line

Tracked down a null flag problem in LoginService.cpp
------------------------------------------------------------------------
r176 | marc | 2009-03-14 18:53:34 -0700 (Sat, 14 Mar 2009) | 2 lines

Added the start of a base service class and started building on that for BRASS based services.
Fixed custom flag editor to save its flags via the new LoginService class that is derived from the BRASSService class
------------------------------------------------------------------------
r175 | marc | 2009-03-13 16:46:07 -0700 (Fri, 13 Mar 2009) | 1 line

Removed forced palette from the main window, finished seanet import for historical transactions
------------------------------------------------------------------------
r174 | marc | 2009-03-13 13:27:55 -0700 (Fri, 13 Mar 2009) | 2 lines

Fixed the logic for overdue in statement processing.
Added aging data to statement creation, moving it from statement printing.
------------------------------------------------------------------------
r173 | marc | 2009-03-13 12:14:00 -0700 (Fri, 13 Mar 2009) | 1 line

Added more latexEscapeString calls for statement processing, specifically on customer name, address, and header and footer messages

adb/ADB.h
/*
** $Id: ADB.h,v 1.21 2002/09/29 18:43:47 marc Exp $
**
**  ADB.h    - Object Database Routines.
**
**  (C)opyright 1998, 1999 by R. Marc Lewis and Blarg! Online Services, Inc.
**  All Rights Reserved
**
**  Unpublished work.  No portion of this file may be reproduced in whole
**  or in part by any means, electronic or otherwise, without the express
**  written consent of Blarg! Online Services and R. Marc Lewis.
**
*****************************************************************************
**
** The ADB classes comprise three classes which create simple access methods
** for one or more MySQL databases.  The databases can be local or remote.
** The Blarg! triple-DES libraries are supported for data protection if
** they are defined and compiled in.
**
** The three classes are:  ADB, ADBTable and ADBList.
**
**    The ADB class provides simple and direct access to a database, 
**    supporting both queries and commands.
**
**    The ADBTable class provides access to a single table within a class.
**    When given the Table name in a database, it automatically determines
**    all data types and the primary key for the table (provided it is a
**    single column).  All operations perform on a single row of the table
**    after a get() or ins() command has been issued.
**
**    ADBList is derived from the ADBTable class and can be used to work
**    with a query that returns multiple rows (such as reports).  Access
**    to individual rows is supported through the inherited ADBTable class.
**
**
** Requres:  - Blarg DES libraries are required if encrypted columns are 
**             to be supported.  Available from Blarg! Online Services, Inc.
**           - MySQL libraries avalable from TCX DataKonsult AB
**
*****************************************************************************
**
** $Log: ADB.h,v $
** Revision 1.21  2002/09/29 18:43:47  marc
** *** empty log message ***
**
** Revision 1.20  2000/06/12 21:48:07  marc
** Added a new class - ADBLogin.  This is a generic login dialog that can
** be used in any Qt based program that uses ADB for database access.
** Altered the defines for using the Qt exentions to ADB.  To use them, the
** user must now define ADBQT=1 in their Makefiles.
**
** Revision 1.19  2000/05/02 20:12:49  marc
** Added the option to return zero dates as empty strings instead of '0000-00-00'.
**
** Revision 1.18  2000/01/03 18:22:58  marc
** Made the order of the fields in the constructors consistant.
**
** Revision 1.17  1999/12/17 18:04:10  marc
** Added the ability for ADB to log to stderr as well as syslog (the default).
**
** Revision 1.16  1999/10/27 00:56:35  marc
** A test for the CVS log function.
**
** Revision 1.15  1999/10/27 00:55:41  marc
** Added some comments to the headers.
** Added appendStr and getLLong to ADBTable.
**
**
*/

#ifndef ADB_H
#define ADB_H

#include <mysql/mysql.h>

#ifdef ADBQT
#include <QtCore/QDateTime>
#include <QtGui/QDialog>
#include <QtGui/QWidget>
#include <QtGui/QLabel>
#include <QtGui/QPushButton>
#include <Qt3Support/q3textview.h>
#include <QtGui/QLineEdit>
#include <QtGui/QComboBox>
#endif

// As MySQL defines bigint as a long long, we need to define it so we don't
// have data overflows or corrupted data.
#ifndef llong
#define llong long long
#endif
#ifndef ullong
#define ullong unsigned long long
#endif


// For speed we pre-allocate enough space to hold column definitions
// up to this number of definitions.
#define ADB_MAXCOLS     128
#define ADB_MAXCOLWIDTH 128


// Logging/Debugging functions
void    ADBLogMsg(int priority, const char *format, ... );
void    ADBDebugMsg(int level,  const char *format, ... );

// Internal definitions for a MySQL column definition
class ADBColumn 
{
public:
    ADBColumn();
    ~ADBColumn();

    void                clear();
    void                clearData();
    
    int                 define(uint columnNo, MYSQL_FIELD *mField, const char *newData = NULL);

    int                 set(const char *newData, int setBackupAlso = 0, int isEncrypted = 0, int useDefKey = 1);
    int                 set(int newValue);
    int                 set(long newValue);
    int                 set(llong newValue);
    int                 set(float newValue);
    int                 set(double newValue);
    int                 append(const char *appVal);
    #ifdef ADBQT
    int                 set(const QDate newValue);
    int                 set(const QTime newValue);
    int                 set(const QDateTime newValue);
    #endif
    
    int                 setColumnName(const char *newName);
    void                setColumnNumber(uint newNum);
    void                setType(enum_field_types newType);
    void                setPrimaryKey(int newIsPrimary);

    const char          *ColumnName();
    const char          *TableName();
    enum_field_types    DataType();
    uint                ColumnNo();
    int                 PrimaryKey();
    int                 ColumnChanged();
    const char          *Data(int useBackup = 0);
    
    // Data conversion functions.
    int                 toInt(int useBackup = 0);
    long                toLong(int useBackup = 0);
    llong               toLLong(int useBackup = 0);
    float               toFloat(int useBackup = 0);
    
    time_t              toTime_t(int useBackup = 0);
    
    #ifdef ADBQT
    // QDateTime functions.
    const QDateTime     toQDateTime(int useBackup = 0);
    const QDate         toQDate(int useBackup = 0);
    const QTime         toQTime(int useBackup = 0);
    #endif
    
    const char          *insStr();
    
    int                 setEncrypted(int isEncrypted, int useDefKey = 1);
    int                 Encrypted();
    
    void                setDebugLevel(int newDebugLevel);

private:
    void                decryptData(void);
    void                encryptData(void);
    
    char                intColumnName[ADB_MAXCOLWIDTH];
    char                intTableName[ADB_MAXCOLWIDTH];
    enum_field_types    intDataType;
    int                 intColumnNo;

    int                 isPrimaryKey;
    int                 intIsEncrypted;
    int                 intUseDefKey;

    int                 debugLevel;
    
    char                *intData;
    char                *intOldData;
    
    char                *workStr;
};


/*
** DBRow   - A class that is the basic construct for retreiving the data
**           from a query.
**
**           This class has no way of performing a query, just loading
**           a row from the database after a query has been performed.
**           when loadRow is called, it loads it from the current database
**           cursor position.
*/

class ADBRow
{
public:
    ADBRow();
    ~ADBRow();
    
    void    setDebugLevel(int newDebugLevel);
    void    setZeroDatesAsNULL(bool newValue);
    void    clearRow();
    int     loadRow(MYSQL_RES *queryRes);
    
    uint    numColumns();
    
    // Magic time.  Here are the operators that will allow access by either
    // the column number or by column name.
    const char *operator[](unsigned int colNo);
    const char *operator[](int colNo);
    const char *operator[](const char *colName);
    
    ADBColumn  *col(unsigned int colNo);
    ADBColumn  *col(int colNo);
    ADBColumn  *col(const char *colName);
    
    ADBColumn  *primaryKey();
    
protected:
    ADBColumn  *findColumn(unsigned int colNo);
    ADBColumn  *findColumn(const char *colName);
    
private:
    ADBColumn   *columns[ADB_MAXCOLS];

    int         intRowDefined;
    uint        numFields;
    
    int         debugLevel;
    bool        zeroDatesAsNULL;
};


/*
** ADB - Object Database Access class.
*/

class ADB
{

public:
    ADB(
      const char *Name  = NULL,
      const char *User  = NULL,
      const char *Pass  = NULL,
      const char *Host  = NULL
    );
    virtual ~ADB();
    
    static const char *defaultHost();
    static  void       setDefaultHost(const char *GlobHost);
    static const char *defaultDBase();
    static  void       setDefaultDBase(const char *GlobDBase);
    static const char *defaultUser();
    static  void       setDefaultUser(const char *GlobUser);
    static const char *defaultPass();
    static  void       setDefaultPass(const char *GlobPass);

    static  void setDebugLevel(int newDebugLevel);
    static  void useSyslog(bool newVal);
    static  void recordUpdates(bool newVal);
    static  void returnEmptyDatesAsNULL(bool newVal);

    int     Connected(void);

    int     query(const char *format, ... );
    float   sumFloat(const char *format, ... );
    int     getrow(void);
    int     getfield(void);
    
    long    dbcmd(const char *format, ... );
    
    // Note that escapeString uses an internal buffer, so multiple calls
    // will trash the return pointer.
    const char  *escapeString(const char *src, int truncLen = 4096);

    ADBRow      curRow;
    ulong       rowCount;
    
protected:
    MYSQL       MyConn;
    MYSQL       *MySock;
    MYSQL_RES   *queryRes;
    
    int         debugLevel;

    // Once the connection is established, these will hold our connect info.
    char        *DBHost;
    char        *DBUser;
    char        *DBPass;
    char        *DBName;
    
    char        *escWorkStr;
    
    int         connected;
};


class ADBTable  : public ADB
{

public:
    ADBTable(
      const char *Table = NULL,
      const char *Name  = NULL,
      const char *User  = NULL,
      const char *Pass  = NULL,
      const char *Host  = NULL
    );
    virtual ~ADBTable();

    // Unless it is specified in the declaration, this can be used to
    // to set the table name that we will be operating on.
    void    setTableName(const char * tabName);
    
    // Clear out all of the data currently in memory.
    void    clearData(void);
    
    // Data assignment members
    int     setValue(uint colNo,          int val);
    int     setValue(const char *colName, int val);
    int     setValue(uint colNo,          long val);
    int     setValue(const char *colName, long val);
    int     setValue(uint colNo,          llong val);
    int     setValue(const char *colName, llong val);
    int     setValue(uint colNo,          float val);
    int     setValue(const char *colName, float val);
    int     setValue(uint colNo,          double val);
    int     setValue(const char *colName, double val);
    int     setValue(uint colNo,          const char *val);
    int     setValue(const char *colName, const char *val);
    int     appendStr(uint colNo,         const char *val);
    int     appendStr(const char *colName,const char *val);
    #ifdef ADBQT
    int     setValue(uint colNo,          const QDate val);
    int     setValue(const char *colName, const QDate val);
    int     setValue(uint colNo,          const QTime val);
    int     setValue(const char *colName, const QTime val);
    int     setValue(uint colNo,          const QDateTime val);
    int     setValue(const char *colName, const QDateTime val);
    #endif
    
    // Data retrieval members
    int     getInt(uint colNo, int useBackup = 0);
    int     getInt(const char *colName, int useBackup = 0);
    long    getLong(uint colNo, int useBackup = 0);
    long    getLong(const char *colName, int useBackup = 0);
    llong   getLLong(uint colNo, int useBackup = 0);
    llong   getLLong(const char *colName, int useBackup = 0);
    float   getFloat(uint colNo, int useBackup = 0);
    float   getFloat(const char *colName, int useBackup = 0);

    const char *getStr(uint colNo, int useBackup = 0);
    const char *getStr(const char *colName, int useBackup = 0);
    
    time_t              getTime_t(uint colNo, int useBackup = 0);
    time_t              getTime_t(const char *colName, int useBackup = 0);
    #ifdef ADBQT
    const   QDate       getDate(uint colNo, int useBackup = 0);
    const   QDate       getDate(const char *colName, int useBackup = 0);
    const   QTime       getTime(uint colNo, int useBackup = 0);
    const   QTime       getTime(const char *colName, int useBackup = 0);
    const   QDateTime   getDateTime(uint colNo, int useBackup = 0);
    const   QDateTime   getDateTime(const char *colName, int useBackup = 0);
    #endif
    
    // Row retrieval access
    long    get(long keyVal);
    int     get(int  keyVal);
    
    // Row insert/update/delete members.
    long            ins(int autoGet = 1);
    virtual void    postIns(void)                   {};
    
    long            upd(int autoGet = 1);
    virtual void    postUpd(void)                   {};

    int             del(long keyVal = 0);
    virtual void    postDel(void)                   {};
    
    // Misc functions.
    int             setEncryptedColumn(uint colNo, int useDefKey = 1);
    int             setEncryptedColumn(const char *colName, int useDefKey = 1);

protected:
    ADBColumn   *columnDefs[ADB_MAXCOLS];
    uint        numColumns;
    uint        primaryKeyColumn;
    
    uint        getColumnNumber(const char *colName);
    
    char        TableName[256];
};


// ADBList simulates a doubly linked list for accessing the columns from 
// a single table returned by a query.  It provides a convenient method 
// to access a full query on a single table.  If joins are used, the ADB 
// class should be used and the results should be obtained manually.
class ADBList : public ADBTable
{
public:
    ADBList(
      const char *Table,
      const char *Name  = NULL,
      const char *User  = NULL,
      const char *Pass  = NULL,
      const char *Host  = NULL
    );
    ~ADBList();

    // The main entry point.  This loads the selected keys into memory and 
    // loads the first row.
    // Only the where portion of the query should be used when giving the
    // query.
    // 
    // Example.  If the Table were "Customers", and contained a column
    // CustomerID (defined as the primary key), and a column "City", and
    // a list of customers based on a certain city is required, the following
    // query would be used:    "where City = 'Seattle'"
    // Which would populate the linked list with all CustomerID's that matched
    // which could then be traversed with first, last, next, prev.
    //
    // If NULL is specified, or no argument is given, it loads all of them.
    long    getList(const char * listQuery = NULL);
    
    // List traversal functions.  Returns zero ((long) 0) if the row was
    // not found.
    // A simple loop would look like this:
    //
    // ADBList  LDB("Customers");
    // LDB.getList("where City = 'Seattle'");
    // for (long liNo == LDB.first(); liNo != 0; liNo = LDB.next()) {
    //    cout << LDB.getLong("CustomerID") << " " << LDB.getStr("City") << endl;
    // }
    long    first(void);
    long    last(void);
    long    prev(void);
    long    next(void);

private:
    long    *keyList;
    long    totKeys;
    long    curKeyNo;
};


#endif // ifdef ADB_H

